hw9_oop/HW_OOP.py

```diff
@@ -0,0 +1,455 @@
+
+Данный класс реализует интерфейс BiologicalSequence
+Добавьте этому классу один любой метод, подходящий по смыслу к аминокислотной последовательности. Например, метод для нахождения изоэлектрической точки, молекулярного веса и т.д.
+Комментарий по поводу метода NucleicAcidSequence.complement, так как я хочу, чтобы вы сделали его опредедённым образом:
+
+При вызове dna.complement() или условного dna.check_alphabet() должны будут вызываться соответствующие методы из NucleicAcidSequence. При этом, данный метод должен обладать свойством полиморфизма, иначе говоря, внутри complement не надо делать условия а-ля if seuqence_type == "DNA": return self.complement_dna(), это крайне не гибко. Данный метод должен опираться на какой-то общий интерфейс между ДНК и РНК. Создание экземпляров NucleicAcidSequence не подразумевается, поэтому код NucleicAcidSequence("ATGC").complement() не обязан работать, а в идеале должен кидать исключение NotImplementedError при вызове от экземпляра NucleicAcidSequence
+
+Вся сложность задания в том, чтобы правильно организовать код. Если у вас есть повторяющийся код в сестринских классах или родительском и дочернем, значит вы что-то делаете не так.
+
+Маленькое замечание: По-хорошему, между классом BiologicalSequence и классами NucleicAcidSequence и AminoAcidSequence, ещё должен быть класс-прослойка, частично реализующий интерфейс BiologicalSequence, но его писать не обязательно, так как задание и так довольно большое (правда из-за этого у вас неминуемо возникнет повторяющийся код в классах NucleicAcidSequence и AminoAcidSequence)
+
+from abc import ABC, abstractmethod
+​
+class BiologicalSequence(ABC):
+    @abstractmethod
+    def __len__(self):
+        pass
+    
+    @abstractmethod
+    def __str__(self):
+        pass
+    
+    @abstractmethod
+    def __getitem__(self, item):
+        pass
+    
+    @abstractmethod
+    def alphabet_check(self):
+        pass
+    
+class NucleicAcidSequence(BiologicalSequence):
+    def __init__(self, sequence):
+        self.sequence = sequence
+        self.allowed = None 
+        self.complement_dict = None
+        
+    def __len__(self):
+        return len(self.sequence)
+    
+    def __str__(self):
+        return str(self.sequence)
+    
+    def __getitem__(self, slc):
+        return self.sequence[slc]
+    
+    def alphabet_check(self):
+        if self.allowed == None:
+            raise NotImplementedError
+        else:
+            alphabet = set(self.sequence)
+            if alphabet.issubset(self.allowed):
+                return 'Good alphabet'
+            else:
+                raise ValueError('Wrong alphabet')
+        
+    def complement(self):
+        if self.complement == None:
+            raise NotImplementedError
+        else:
+            self.complemented = []
+            for letter in self.sequence:
+                self.complemented.append(self.complement_dict[letter])
+        return ''.join(self.complemented)
```

Comment:
При комплементировании мы ожидаем получить данные такого же типа, а не строку. Чинится это вот так

```suggestion
        return type(self)(''.join(self.complemented))
```

При комплементировании ДНК мы получим ДНК и так для любого типа